### PR TITLE
Portal marketmaker

### PR DESCRIPTION
**Pull request by kahuang**

### CODE REVIEW

File: packages/portal-contracts/src/Portal.sol [hunk 1]
```diff
@@ -80,11 +80,11 @@ struct Escrow {
     /**
      * @dev If correct amount is paid to script hash, who keeps the escrow?
      */
-    address successOpenEscrow;
+    address successRecipient;
```
Comment (dcposch) [inline packages/portal-contracts/src/Portal.sol:83]: this fixes https://github.com/dcposch/silverportal/pull/28/files , which looks like it find-replaced `Recipient` with `OpenEscrow` everywhere?

File: packages/portal-marketmaker/src/index.js [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
...veSettlement","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"contract IBtcTxVerifier","name":"_btcVerifier","type":"address"}],"name":"setBtcVerifier","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"_minConfirmations","type":"uint256"}],"name":"setMinConfirmations","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"newOwner","type":"address"}],"name":"setOwner","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"_stakePercent","type":"uint256"}],"name":"setStakePercent","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"escrowID","type":"uint256"}],"name":"slash","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"stakePercent","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"token","outputs":[{"internalType":"contract IERC20","name":"","type":"address"}],"stateMutability":"view","type":"function"}]`;
```
Comment (dcposch) [inline packages/portal-marketmaker/src/index.js:69]: check out how `portal-web` uses typechain

this lets you avoid pasting in ABIs = easier to update when contracts change

also provides a typed API

File: packages/portal-marketmaker/src/index.js [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+	destScriptHash
+	deadline
+	id
+}_meta { block { number }}}`;
+
+const myBtcAddress = "mqHEJi3boWQHwo2x8fEpXejyDWj37yTtDG";
+const myBtcKey = "cNdotpEDrvEhCp7rABaent5cqnvrX9gGph3eL1PPQLt3AxmAwETh";
+//const myBtcAddress = "0x529ccdd3112490bc59754892787e93f124520e78";
+
+var bidSats = BigInt(0);
```
Comment (dcposch) [inline packages/portal-marketmaker/src/index.js:57]: give these globals one-liner comments

is `bidSats` the total amount of all open bids?

if a bid gets partially filled, does bidSats go down by that amount?

File: packages/portal-marketmaker/src/index.js [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+const portal = new ethers.Contract(portalAddress, portalJSONABI, provider);
+const portalWithSigner = portal.connect(signer);
+
+const wbtc = new ethers.Contract(wbtcAddress, wbtcAbi, provider);
+const wbtcWithSigner = wbtc.connect(signer);
+
+
+var minblock = 0;
+
+function handlePendingOrder(order) {
```
Comment (dcposch) [inline packages/portal-marketmaker/src/index.js:86]: i'd recommend giving these functions one-liners too

for this one, if i'm understanding it right, something like

```
// Adds order we just placed to our liquidity totals [bid/ask][Sats/Liquidity] 
```

File: packages/portal-marketmaker/src/index.js [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+		bidSats += amountSats
+		bidLiquidity += amountSats * priceTokPerSat
+	} else {
+		askSats += amountSats
+		askLiquidity += amountSats * priceTokPerSat
+	}
+}
+
+async function loadPendingOrders() {
+	await axios.post('https://api.thegraph.com/subgraphs/name/kahuang/silver-portal', JSON.stringify({
```
Comment (dcposch) [inline packages/portal-marketmaker/src/index.js]: thegraph API endpoint used in a few places, recommend pulling to a constant

File: packages/portal-marketmaker/src/index.js [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+}
+
+async function loadWBTCBalance() {
+	const myWbtcBalance = (await wbtc.balanceOf(myAddress)).toBigInt();
+	myWbtcSats = myWbtcBalance * BigInt(satsToBTC); 
+}
+
+async function loadBTCBalance() {
+	const utxos = await axios.get(
+		`https://sochain.com/api/v2/get_tx_unspent/${sochain_network}/${sourceAddress}`
```
Comment (dcposch) [inline packages/portal-marketmaker/src/index.js:226]: using this provider bc getblock doesn't support what we need?

File: packages/portal-marketmaker/src/index.js [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+	return new Promise((resolve) => {
+	  setTimeout(resolve, ms);
+	});
+}
+
+async function handlePendingEscrows(escrows) {
+	promises = [];
+	for (var i = 0; i < escrows.length; i ++ ) {
+		escrow = escrows[i];
+		console.log("Handline escrow: ", escrow);
```
Comment (dcposch) [inline packages/portal-marketmaker/src/index.js]: nit, spelling

File: packages/portal-marketmaker/src/index.js [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+		var buf = Buffer.from(strip0x(escrow.destScriptHash), "hex");
+		var addr = bitcore.Address.fromScriptHash(buf, 'testnet', 'scripthash').toString();
+		const amountSatsDue = parseInt(escrow.amountSatsDue);
+		console.log("Sending ", amountSatsDue, " sats to ", addr);
+		const txid = (await sendBitcoin(addr, amountSatsDue)).txid;
```
Comment (dcposch) [inline packages/portal-marketmaker/src/index.js:126]: so... does `sendBitcoin` return a promise that waits for block confirmations? will resolve veeery slowly

Comment (kahuang) [reply packages/portal-marketmaker/src/index.js:126]: sendBitcoin sends the tx to the mempool and returns the transaction id. we wait for confirmations in the next part

**Review by dcposch — COMMENTED** [review]
nice prototype! for testnet all we need is something that lets people trade with Silver Portal while we're away, and i think this accomplishes that 🚀

in prod we can make a thin market initially

pro market makers will write their own bot, but def useful for them to have a reference implementation

--

the main thing i'd fix about this one before mainnet is that if you restart the MM, i think it'll double send Bitcoin transactions

to avoid that, rather than a loop like

### Current

(Loop)
1. Get open escrow
2. Send Bitcoin, waiting for confirmations
3. Prove payment

### Proposed

MM is a short-lived process, runs every few minutes.

- Load open escrows.
- Load pending Bitcoin transactions (from the mempool)
- Load recent Bitcoin transactions to/from our addresses
- If we've sent any transactions that close an open escrow, with enough confirmations, call `proveSettlement` to close.
- If there are any open escrows that DON'T already have a pending (mempool) or recently mined transaction, then send a Bitcoin transaction.

+ For safety, I'd save the escrowID immediately before sending the Bitcoin tx, say to sqlite. The logic above *shouldn't* ever try to close the same escrow twice, but worth checking the local DB and crashing with a descriptive error if that escrowID is already recorded as having been sent.

This is similar to the BtcMirror submitter--stateless (minus the safety DB described at the end), short lived (completes in ~30 seconds), runs on some cron-like mechanism.

**Review by dcposch — COMMENTED** [review]
👍

**Pull request merged**